### ISSUE TITLE
rclone http listener

### ISSUE THREAD

**Issue by remusb** [ncw/rclone#2111]
I was thinking about how to give _cache_ an easy way to purge specific paths or objects from its storage. Plus all the work around polling with _drive_ gave me an idea on how to do this better. 

What would you think about adding an http listener which can receive requests and proxy them to the remotes that subscribe to it?

Some of the things I thought about:
- would start up on first subscription from a remote (similar to atexit) 
- remotes can subscribe with a callback and specify parameters to compound the URL for it
- the http listener will transform GET query parameters to a simple map[string]string
- .... when there's a POST or PUT it will callback with a Reader for the body
- try to keep it lightweight, get the request and pass it down to the subscribed remote
- could open a way for cloud providers to push notifications to rclone directly rather than polling (but there are complications for this use case)

My usage would first be to have a cache purge command that checks if there's a rclone process listening and push a request to purge something to the long running rclone process. More experienced folks could curl rclone directly :)

Is it a niche topic? I know cache is probably the only remote that has the limitation of running in parallel (bolt db lock).

**Comment by naeloob** [comment]
I like the Idea!!!
I was thinking about something like that in order to change parameters on the fly, without having to stop and start again rclone. Just like a signal but from http, that way it could work on every SO and have more flexibility and options than just a signal.

**Comment by daniel-loader** [comment]
https://forum.rclone.org/t/cache-ability-to-purge-directory-cache-without-purging-cached-files/4839

May be related content to add to this discussion.

**Comment by daniel-loader** [comment]
To go a little further, today I've noticed some old entries in the upload cache directory that don't seem to have been deleted, that is to say, files not directories - I confirmed they were on the remote via the webUI and stopped the mount and deleted the cache upload files and remounted but wondering what happened.

I'd add it'd be great to be able to do one off force uploads via webhooks. Given there's a lot of potential things to trigger or alter during runtime, using SIGHUPs and similar was going to be a losing battle.

**Comment by remusb** [comment]
@daniel-loader 
The Sonarr/Radarr issues (or better said, cache expiration related) should be improved after https://github.com/ncw/rclone/pull/2094 gets in as long as the cloud provider is drive or amazon as those are using change notifications.

But sure, giving the ability to other tools to expire cache on their demand will make things work better together

**Comment by daniel-loader** [comment]
That's great, I'll give it a go once it hits beta.

Can you give a tiny bit more detail as to what the difference would look like from the end user POV so I can test it before/after?

**Comment by ncw** [comment]
I have made a proposal for discussion with some ideas that I've had floating about for a while!

The [overview docs](https://github.com/ncw/rclone/blob/rc/docs/content/rc.md) give an explanation of how things fit together.  The [rclone rc](https://github.com/ncw/rclone/blob/rc/cmd/rc/rc.go) command describes the rclone client for the remote control.  You can use curl too if you want.

How to use the internal API is [described in the rc.go](https://github.com/ncw/rclone/blob/rc/fs/rc/rc.go) and there are some [examples of simple APIs](https://github.com/ncw/rclone/blob/rc/fs/rc/internal.go) which I implemented to get the ball rolling!

It is strictly an RPC style interface so no streaming or web pages.  The latter would be relatively easy to add though.

@remusb let me know what you think - is this sufficient for your use?

Any comments gratefully received!

**Comment by remusb** [comment]
@ncw oh wow, you move too fast. I was still at starting the refactorization of the PR I did.
This all looks great. A couple of questions:

1. do we want other commands than `mount` to start up on `--rc`? I can only think of sync being useful if it's done on expected long running syncs
2. seems to be ready for use. I'll open a new PR to add this in cache

**Comment by ncw** [comment]
> do we want other commands than mount to start up on --rc? I can only think of sync being useful if it's done on expected long running syncs

I think potentially any rclone command could use it.  sync, copy, move, copyto, moveto are all good candidates.  I don't see a disadvantage for enabling it for all commands particularly.

**Comment by remusb** [comment]
This got in master and it's available as a beta: https://beta.rclone.org/v1.39-241-g67797164/ (should be complete soon)

Please read the docs about RC mode and cache support for it:
- https://github.com/ncw/rclone/blob/master/docs/content/rc.md
- https://github.com/ncw/rclone/blob/master/docs/content/cache.md#cache-and-remote-control---rc

**Comment by seuffert** [comment]
One thing I noticed while testing:

If you are using rclone with cache & and crypt (gd/cache/crypt) and you want to invalidate a file/dir though the new rc api, you must use use the encrypted filenames. This really makes things complicated from a user perspective.

**Comment by seuffert** [comment]
Oh and one more little detail:
Invalidating a directory via the rc api only invalidates the cache remote DB, the mount dir cache will still have the old entries until dir cache time expires... I bet there will be some people scratching there heads about this 😉

**Comment by remusb** [comment]
Both options are supported in the latest master @seuffert 

I realized another problem will be absolute paths in the case of a mount. They won't work as the backends will not have the context of where they're running in and how. Paths will always have to be relative from the mount point.
This isn't something that can be changed though at this point.

**Comment by seuffert** [comment]
@remusb great, thx! :) will test it later
i added one comment on your commit

**Comment by daniel-loader** [comment]
Can anyone help me work out how to use rc to purge a directory? An example or two would be great!

**Comment by seuffert** [comment]
example to expire a relative sub-folder in mount:
` rclone rc cache/expire remote=path/to/sub/folder/`

**Comment by daniel-loader** [comment]
Thanks, I'll go try it out now. 
Any example for the curl request alternative? 

Once I get it all sorted I'll do a PR to update the documentation page a bit more.

**Comment by seuffert** [comment]
i updated the above example to include curl

**Comment by daniel-loader** [comment]
How does one dictate which cached remote is being purged? The curl string sent doesn't indicate which remote it is?

**Comment by seuffert** [comment]
the port can not be shared, every rclone mount must run on its own port number

**Comment by daniel-loader** [comment]
Ah right, as I left it default rather than dictating a port on the http listener I hadn't considered the issue. Glad to have it cleared up!

**Comment by seuffert** [comment]
from the [rclone rc](https://rclone.org/commands/rclone_rc/) documentation:
`--url string   URL to connect to rclone remote control. (default "http://localhost:5572/")`
